### PR TITLE
gNMI config subscription extension documentation

### PR DESCRIPTION
**Pull request by karimra**
Documentation page to complement https://github.com/openconfig/gnmi/pull/169

### CODE REVIEW

**Comment by karimra** [comment]
@robshakir @dplore, would you mind taking a look at this PR?

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+1) The client subscribes to path P1 with the `ConfigSubscription` extension
+present with the action `ConfigSubscriptionStart`.
+2) The server processes the subscription request as usual but will only send
+updates for the configuration schema nodes under the path P1.
+3) The client sends a Set RPC with the configuration changes to the path P1
+**without** the `CommitConfirm` extension.
+4) The server processes the Set RPC as usual and sends the updates for the
+configuration schema nodes under the path P1.
+5) As all the configuration updates are sent, the server sends the
```
Comment (dplore) [inline rpc/gnmi/gnmi-config-subscriptions.md:132]:
```suggestion
5) After all the configuration updates are sent, the server sends the
```

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+}
+```
+
+## 2.2 Actions
+
+### 2.2.1 ConfigSubscriptionStart
+
+A `ConfigSubscriptionStart` message is used by a gNMI client in a
+`SubscribeRequest` to indicate that it wants to start a ConfigSubscription.
+The target must respond exclusively with configuration data relevant to the
```
Comment (dplore) [inline rpc/gnmi/gnmi-config-subscriptions.md:93]: I think it would be good to include an explicit definition if a notification is expected to be sent with a subscribe response for P1.  Following the pattern of subscribe, I am thinking the target should send a subscribe response with the path(s) requested, even in an ON_CHANGE subscription.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+1) The client subscribes to path P1 with the `ConfigSubscription` extension
+present with the action `ConfigSubscriptionStart`.
+2) The server processes the subscription request as usual but will only send
+updates for the configuration schema nodes under the path P1.
+3) The client sends a Set RPC with the configuration changes to the path P1
+and **with** the `CommitConfirm` extension present.
+4) The server processes the Set RPC as usual and sends the updates for the
+configuration schema nodes under the path P1.
+5) As all the configuration updates are sent, the server sends the
```
Comment (dplore) [inline rpc/gnmi/gnmi-config-subscriptions.md:167]:
```suggestion
5) After all the configuration updates are sent, the server sends the
```

**Review by dplore — COMMENTED** [review]
A few minor comments for you

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 2 above)
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:93]: I added the following paragraph for clarification:

The target must respond with an initial set of updates followed by a
`SubscribeResponse` where the `sync_update` field is set to `true`.
However, if the `updates_only` field in the `SubscribeRequest` was set to
`true`, the target should skip sending the initial updates and only send
a `SubscribeResponse` with `sync_update` set to `true` as per the base gNMI
specification.

Comment (dplore) [reply rpc/gnmi/gnmi-config-subscriptions.md:93]: thank you!

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+The `ConfigSubscriptionSyncDone` message includes two optional fields:
+
+* `commit_confirm_id`: A commit confirm ID assigned by the client which
+initiated the commit.
+The commit can be initiated via gNMI (using the CommitConfirmed Extension),
+NETCONF, or any other management interface. Applicable only if the commit
+confirmed option is used.
+* `server_commit_id`: An optional internal ID assigned by the target.
+
+In the case a commit happens before the `sync_response: true` and the server
```
Comment (robshakir) [inline rpc/gnmi/gnmi-config-subscriptions.md:121]: Adding a second thread here -- what happens if we have two configuration changes -- i.e.:

* client A makes a change
* client B makes a change
* client C has a `ConfigSubscription` open

It seems like, in this case, `sync_response: true` might be returned at any point in the lifetime of the subscription (when whatever existed in the data tree at the time client C created its subscription). We want `ConfigSubscriptionSyncDone` to be returned at the point that client A and client B have had their changes "committed" (let's clarify semantics here) -- this is rather independent to when `sync_response: true` is sent.

Another implication here is that we cannot tell the exact provenance of an update within the `Subscribe` stream if there are changes being made that do not block on the "completion" of a previous change.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+updates for the configuration schema nodes under the path P1.
+3) The client sends a Set RPC with the configuration changes to the path P1
+**without** the `CommitConfirm` extension.
+4) The server processes the Set RPC as usual and sends the updates for the
+configuration schema nodes under the path P1.
+5) After all the configuration updates are sent, the server sends the
+`ConfigSubscriptionSyncDone` message to the client in a SubscribeResponse
+message.
+This message does not contain a `commit_confirmed_id` and may contain a
+`server_commit_id`
```
Comment (robshakir) [inline rpc/gnmi/gnmi-config-subscriptions.md:142]: Maybe we should just have a "done" bool in this message as well as these IDs -- otherwise we have to do a check for a non-nil message, right?

**Review by robshakir — COMMENTED** [review]
Thanks for the specification doc. It's very useful to understand this further.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+**Contributors:** Roman Dodin, Matthew MacDonald
+
+**Date:** July 29th, 2024
+
+**Version:** 0.1.0
+
+## 1. Purpose
+
+Performing configuration management and handling configuration drift is
```
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:11]: That's correct. That change can be reverted by the management system or "accepted" and be made part of the expected configuration of the device.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+## 2.2 Actions
+
+### 2.2.1 ConfigSubscriptionStart
+
+A `ConfigSubscriptionStart` message is used by a gNMI client in a
+`SubscribeRequest` to indicate that it wants to start a ConfigSubscription.
+The target must respond exclusively with configuration data relevant to the
+created subscription.
+The target must respond with an initial set of updates followed by a
+`SubscribeResponse` where the `sync_update` field is set to `true`.
```
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:97]: It is the `sync_update` from the base specification. Essentially this means no change to the base spec.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+However, if the `updates_only` field in the `SubscribeRequest` was set to
+`true`, the target should skip sending the initial updates and only send
+a `SubscribeResponse` with `sync_update` set to `true` as per the base gNMI
```
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:100]: Correct, no changes to the base spec. I might have not expressed it clearly enough.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+specification.
+
+### 2.2.2 ConfigSubscriptionSyncDone
+
+A `ConfigSubscriptionSyncDone` message is used by a gNMI target in a
+`SubscribeResponse` to indicate a commit boundary to the client.
+A commit boundary marks the completion point of a specific set of
```
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:107]: Yes, I can rephrase it to `A commit boundary marks the completion point of a single set of changes associated with a commit`

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:97]: Ah, you mean `sync_response`?

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:122]: > is the implication here that we should hold back this sync_reponse: true until all the leaves that are the result of a new commit are sent?

It means that we should hold back the `ConfigSubscriptionSyncDone` until after `sync_reponse: true`.
Essentially, do not send any `ConfigSubscriptionSyncDone` until the initial updates are sent, even if all updates related to a commit were already streamed to the collector.

Example:

- A clientA creates a subscription with this extension to root `/`
- While the initial updates are still being streamed to clientA, clientB changes the hostname from `dut1` to `dut2`
- ClientA might receive the hostname twice (`dut1` then `dut2`) but would only receive a `ConfigSubscriptionSyncDone` after all the initial updates are streamed and a `sync_reponse: true` is sent.

File: rpc/gnmi/gnmi-config-subscriptions.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+configuration changes.
+It indicates that all changes within that set have been committed and
```
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:109]: Committed means the interface (gNMI, netconf, CLI,...) that was used to change the values returned a nil/zero error.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 5 above)
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:143]: `3.1 Configuration changes without Commit Confirmed` is the case the change is made without CommitConfirmed and the device does not provide a `server_commit_id`. 
Checking a nil msg is equivalent to checking for a boolean. In both cases, the client has to check for a nil message.

Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:143]: I don't think that's necessarily true -- if I can do `foo.GetMessage().GetBool()` then I don't have to do a nil-check for the `message` field in the Go API at least. However, regardless, my view is that having explicit fields makes debugging much easier, since seeing `done: true` is much clearer than having to remember that an empty message in a `oneof` means something :-)

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 4 above)
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:122]: the config store does not need to support parallel commits for this situation to happen.

assume we take N seconds to stream all updates corresponding to a config change.

t=0 - config change 1 happens and is finished
t=1 - streaming for updates from config change 1 starts
t=Y (Y<N) - config change 2 happens and is finished
t=Y+1 - streaming for updates from config change 2 starts
t=N - streaming for updates from config change 1 ends
t=Y+N - streaming for updates from config change 2 ends

in this scenario -- it's the streaming of the updates that overlaps, not the commits. 

when the client receives the `ConfigSubscriptionSyncDone` with the commit ID of config change 1 in it, it may already have received updates for paths that were impacted only by config change 2 -- so the semantic that this marker means "you have the state that reflects the state only after change 1" is broken.

the only way to avoid this is by saying that the config store locks until all updates are streamed (i.e., N seconds) (which seems bad for performance), or there is a strict send buffer with no coalescing possible (which I don't think is the right behaviour -- and is already broken by existing gnmi implementations).

why does coalescing matter? well, if both change 1 and change 2 change `/foo/bar` and the update for `/foo/bar` hasn't been sent, then it's acceptable for an gNMI server to coalesce the two, and only send the value for `/foo/bar` that corresponds to the value after the second change.

i think the semantics that you're trying to define here for `ConfigSubscriptionSyncDone` are problematic with any kind of concurrency. to me, it's better to say that the gNMI server gives looser semantics about what it can guarantee -- i.e., it guarantees only that it is "done" with processing the changes that are covered by a particular commit. if the client *really* wants to ensure that its config subscription reflects exactly that intended state, it can implement locking between changes without adding complexity at the server side.

[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:122]: this would be worth clarifying in the text -- since (as above) I read it the other way around. I think simply saying that `ConfigSubscriptionSyncDone` cannot be sent until `sync_response: true` has been sent is reasonable. (this will still need some per-client state to be stored about whether the initial sync has been completed -- but at least does not mean that the `sync_response` behaviour is changed.)

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:109]: let's clarify this in the text.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:107]: sgtm.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:100]: I'd maybe just preface the paragraph with this statement -- there is no change to the base behaviour of the `Subscribe` rpc.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md:11]: perhaps worth clarifying -- since i'm not sure it is a widely understood term (or at leas it isn't for me :-))

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:97]: My bad, fixing

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:122]: The paragraph is more about allowing a server to send the committed changes updates before the `sync_response: true`. Meaning that updates do not have to be buffered but the `ConfigSubscriptionSyncDone` has to be.
Rephrasing to this:

```
In the case a commit happens before the `sync_response: true` the server
cannot send a `ConfigSubscriptionSyncDone` until the `sync_response: true` has
been sent. The server may send the committed changes updates before the
`sync_response: true`.
```

Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:122]: I think this definition works: `it guarantees only that it is "done" with processing the changes that are covered by a particular commit.`. The streaming of changes depends on whether they were coalesced.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: added:
```
Performing configuration management and handling configuration drift is one of
the main features of a higher-level management system or orchestrator.
Configuration drift occurs when a device's configuration changes independently
of the expected configuration defined by the management system. Such changes
may arise from manual interventions, automated processes outside the management
system, or unexpected behavior. When drift is detected, the management system
can either revert the configuration back to the expected configuration or
incorporate the change by updating its records to include the new configuration
as the expected configuration.
```

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: added:
```
The base behavior of the `Subscribe` RPC remains unchanged: the target must
respond with an initial set of updates, followed by a `SubscribeResponse`
with the `sync_response` field set to true. However, if the updates_only
field in the SubscribeRequest is set to true, the target should omit the
initial updates and instead send only a SubscribeResponse with `sync_response`
set to true, in accordance with the gNMI specification.
```

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: done

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: added:

```
It indicates that all changes within that set have been committed -- meaning
the mechanism used to apply the changes reported no errors -- and
that all notifications triggered by the commit have been streamed to
the client.
```

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 5 above)
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md:150]: agree, will update the proto

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md]: LGTM, thank you.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md]: Great, thanks.

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md]: It seems like the fact that the 'commit' semantics are defined by the definition per interface -- so can we say "gNMI, NETCONF, CLI etc., returned a non-zero value" explicitly?

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md]: Thanks, LGTM.

Comment (robshakir) [reply rpc/gnmi/gnmi-config-subscriptions.md]: ACK -- suggest to update the text here to describe that this is the semantic meaning of receiving a "done". I might also suggest adding the example of concurrency that means that this design decision is required in an appendix for future readers.

Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: Updated and added appendix

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: Done

File: rpc/gnmi/gnmi-config-subscriptions.md (hunk 4 above)
Comment (karimra) [reply rpc/gnmi/gnmi-config-subscriptions.md]: @robshakir @dplore could you please have a look at the update?

**Review by robshakir — APPROVED** [review]
